compiler/src/iree/compiler/Dialect/HAL/IR/HALTypes.cpp

```diff
@@ -308,6 +308,128 @@ SmallVector<IREE::HAL::DeviceTargetAttr, 4> DeviceTargetAttr::lookup(
   return {};  // No devices found; let caller decide what to do.
 }
 
+// Returns a set of all configuration attributes from all device targets with
+// a configuration set. Targets with no configuration set are ignored.
+static SmallVector<DictionaryAttr> lookupOptionalConfigAttrs(Operation *op) {
+  SmallVector<DictionaryAttr> configAttrs;
+  auto targetAttrs = IREE::HAL::DeviceTargetAttr::lookup(op);
+  for (auto targetAttr : targetAttrs) {
+    auto configAttr = targetAttr.getConfiguration();
+    if (configAttr) configAttrs.push_back(configAttr);
+  }
+  return configAttrs;
+}
+
+// Returns a set of all configuration attributes from all device targets.
+// Returns nullopt if any target is missing a configuration attribute.
+static std::optional<SmallVector<DictionaryAttr>> lookupRequiredConfigAttrs(
+    Operation *op) {
+  SmallVector<DictionaryAttr> configAttrs;
+  auto targetAttrs = IREE::HAL::DeviceTargetAttr::lookup(op);
```

Comment:
Do these need to be on op, or is it something where it would walk up to parent?

Reply:
lookup does the walking so if we were to scope functions or other kinds of regions with `hal.device.targets` attrs we'd get those first/instead of any ancestor ones